impl/src/main/java/io/jsonwebtoken/impl/DefaultJwtParser.java

```diff
@@ -293,7 +303,7 @@ public Jwt parse(String jwt) throws ExpiredJwtException, MalformedJwtException,
             base64UrlEncodedDigest = sb.toString();
         }
 
-        if (base64UrlEncodedPayload == null) {
+        if (base64UrlEncodedPayload == null && payloadRequired) {
```

Comment:
I was thinking about triggering exceptions due to `payloadRequired()` late last night and how an exception might manifest for the user.

Let's say someone configures `requireSubject("foo")` and then there isn't a payload.  The exception they'll see is a generic `MalformedJwtException` with a message `"JWT string '<whatever>' is missing a body/payload"`, and doesn't explain at all _why_ the JWT is required to have a payload.

To me, this seems to be undesirable.  If you think about it, in practice, almost no one really cares that there is a payload or not.  What they *really* care about is if something *in* the payload exists and/or matches an expectation.  And we already have support for enforcing and communicating those things via the other `require`* methods in the class.

(Also FWIW, I don't think `MalformedJwtException` is correct here - the JWT _isn't_ malformed - it's just that expectation/validation requirements have failed, so I think we'd have to throw another exception - probably `UnsupportedJwtException`, and that's if we throw at all...)

Which brings me to my next point:  do we need this check at all?

What if we didn't have a `payloadRequired` check at all and just let the empty string default as done on line 330 in this PR below?  The other `require*` methods would still fail with a meaningful/descriptive exception as is more desirable.

Would there be any problems?  Would any of the existing many tests in the project fail?  That would tell us a lot.

Reply:
Good point. 'Malformed' is certainly not the correct wording in that case. It's more an unmet requirement the callers asked for themselves.
All I can say from during the implementation - no tests fail if this check just goes away. The only thing that needs to be added would be the `isEmpty()` check for the payload below.

Reply:
Ok - in this case, I don't think we need the concept of `payloadRequired` at all and just keep the `isEmpty()` check.